labs/cryptol-overview/overview.md

```diff
@@ -0,0 +1,544 @@
+[0x31, 0x32]
+```
+ * **Records** - Records allow for more complex data structures to be formed, where fields may be accessed by their tnames.
+ 
+```haskell
+Cryptol> let point = {x = 10:[32], y = 25:[32]}
+Cryptol> point.x
+0x0000000a
+Cryptol> point.y
+0x00000019
+```
+
+## Operators
+
+Cryptol provides a collection of built in operators to build expressions and perform computations. Some familiar operators which appear in cryptographic applications include:
+
+* `~` -- bitwise inversion
+* `&&`, `||`, `^` -- bitwise logical operations
+* `==`, `!=` -- structural comparison
+* `>=`, `>`, `<=`, `<` -- nonnegative word comparisons
+* `+`, `-`, `*`, `/`, `%`, `**` -- wordwise modular arithmetic
+* `>>`, `<<`, `>>>`, `<<<` -- shifts and rotates
+* `#` -- concatenation
+
+There are many more. A list of the currently defined symbols and operators are available by running `:browse` in the interpreter.
+
+An interesting feature of Cryptol's type system is that operators are typed and which you can check with the `:t` command in the interpreter just as you can for data:
+
+```haskell
+Cryptol> :t (&&)
+(&&) : {a} (Logic a) => a -> a -> a
+```
+
+In a nutshell, this indicates that the bitwise and operater (`&&`) operates on two elements of type `a` that are in the "`Logic`" typeclass and returns another element of the same type. Sometimes this information is useful for debugging a design.
+
+## Primitives
+
+Cryptol offers a robust set of primitives to construct more complex functions and specifications with. The built-in primitives cover a wide variety of purposes. A sampling of these built-ins include:
+
+ * **List manipulation** -- `take`, `drop`
+ 
+```haskell
+Main> take `{1} [1, 2, 3]
+[1]
+Main> drop `{1} [1, 2, 3]
+[2, 3]
+```
+ 
+ * **Structural manipulation** -- `split`, `groupBy`
+
+```haskell
+Main> split `{3} [1, 2, 3, 4, 5, 6]
+[[1, 2], [3, 4], [5, 6]]
+Main> groupBy `{3} [1, 2, 3, 4, 5, 6]
+[[1, 2, 3], [4, 5, 6]]
+```
+
+ * **Arithmetic** -- `sum`, `min`, `max`
+
+```haskell
+Main> sum [1, 2, 3, 4, 5]
+15
+Main> min 5 10
+5
+Main> max 5 10
+10
+```
+
+Inside of the Cryptol interpreter you can inspect the symbols, functions, and primitives that are defined within the current context using the `:browse` (or simply `:b`) command. Doing so at the beginning of a fresh interpreter session will show you the list of built-ins and other information about the environment:
+
+```haskell
+Main> :b
+    all : {n, a} (fin n) => (a -> Bit) -> [n]a -> Bit
+    and : {n} (fin n) => [n] -> Bit
+    any : {n, a} (fin n) => (a -> Bit) -> [n]a -> Bit
+    ...
+    zext : {m, n} (fin m, m >= n) => [n] -> [m]
+    zip : {n, a, b} [n]a -> [n]b -> [n](a, b)
+    zipWith : {n, a, b, c} (a -> b -> c) -> [n]a -> [n]b -> [n]c
+```
+
+
+## Functions
+
+Cryptol allows users to define functions allowing subcomponents to be reused or allow greater clarity in a program. Functions are typically specified in two parts: the *type definition* and the *function definition*.
```

Comment:
allow 3x. How about:

Cryptol enables users to define functions, thereby allowing subcomponents to be reused and providing greater clarity.

Or just leave the whole sentence out. What programming language doesn’t allow functions?

Reply:
This was part of a larger conversation on one of the slides I was working from. But one of the examples doesn't work the same so this section came out a little light. I think it works in the flow of things here and if people skim over it, I don't think it's a big deal.